File: solution/187/solution.py
class Solution(object):
    def findRepeatedDnaSequences(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        result = []
        if len(s) < 10:return result
        charlist = ['A', 'C', 'G', 'T']
        temp = 0
        hash_map = dict()
        for i in xrange(9, -1, -1):
            temp = (temp << 2) + charlist.index(s[i])
        hash_map[temp] = 1
        for x in s[10:]:
            temp = (temp >> 2) | (charlist.index(x) << 18)
            hash_map[temp] = hash_map.get(temp, 0) + 1
        for k, v in hash_map.items():
            if v > 1:
                getstr = ''
                data = k
                for _ in xrange(10):
                    getstr += charlist[data & 3]
                    data = data >> 2
                result.append(getstr)
        return result


s = Solution()
print s.findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT")

File: solution/187/solution.cpp
class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        vector<string> result;
        if(s.length()<10)return result;
        int hash_map[1048576] = {0};
        char listchar[] = {'A', 'C', 'G', 'T'};
        int charmap[128] = {0};
        charmap['A'] = 0;
        charmap['C'] = 1;
        charmap['G'] = 2;
        charmap['T'] = 3;
        int num = 0;
        for(int i = 9;i>=0;i--){
            num = (num<<2) + charmap[s[i]];
        }
        hash_map[num] ++;
        for(int i=10;i<s.length();i++){
            num = num >> 2 | charmap[s[i]] << 18;
            hash_map[num] ++;
        }
        for(int i=0;i<1048576;i++){
            if(hash_map[i]>1){
                int data = i;
                string str;
                for(int j=0;j<10;j++){
                    str += listchar[data&3];
                    data=data>>2;
                }
                result.push_back(str);
            }
        }
        return result;
    }
};